File: old kit/todo-system-php/includes/GoogleSheetsIntegration.php
<?php
require_once 'config.php';
require_once 'TodoManager.php';

/**
 * Google Sheets Integration for Todo Management
 * Syncs todos with Google Sheets (CSV-based for simplicity)
 */
class GoogleSheetsIntegration {
    private $todoManager;
    private $csvFile;
    
    public function __construct($csvFile = null) {
        $this->todoManager = new TodoManager();
        $this->csvFile = $csvFile ?? Config::get('data_dir') . 'google_sheets_todos.csv';
        $this->initializeCSV();
    }
    
    /**
     * Initialize CSV file with headers
     */
    private function initializeCSV() {
        if (!file_exists($this->csvFile)) {
            $headers = ['ID', 'Todo Item', 'Status', 'Date Added', 'Due Date', 'Priority', 'Notes', 'Source'];
            
            $fp = fopen($this->csvFile, 'w');
            fputcsv($fp, $headers);
            fclose($fp);
            
            Config::log("Initialized Google Sheets CSV file: {$this->csvFile}");
        }
    }
    
    /**
     * Add todo to Google Sheets (CSV)
     */
    public function addTodo($description, $dueDate = '', $priority = 'Medium', $notes = '', $source = 'sheets') {
        $todoId = $this->getNextId();
        $dateAdded = date('Y-m-d');
        
        $row = [
            $todoId,
            $description,
            'Pending',
            $dateAdded,
            $dueDate,
            $priority,
            $notes,
            $source
        ];
        
        $fp = fopen($this->csvFile, 'a');
        fputcsv($fp, $row);
        fclose($fp);
        
        // Also add to main database
        $sourceData = [
            'due_date' => $dueDate,
            'notes' => $notes,
            'channel' => 'sheets'
        ];
        
        $dbTodoId = $this->todoManager->addTodo($description, 'sheets', strtolower($priority), $sourceData);
        
        Config::log("Added todo to Google Sheets: #$todoId - $description");
        return $todoId;
    }
    
    /**
     * Get next available ID from CSV
     */
    private function getNextId() {
        if (!file_exists($this->csvFile)) {
            return 1;
        }
        
        $lines = file($this->csvFile, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES);
        
        if (count($lines) <= 1) { // Only header or empty
            return 1;
        }
        
        // Get the last line and extract ID
        $lastLine = end($lines);
        $data = str_getcsv($lastLine);
        
        if (!empty($data[0]) && is_numeric($data[0])) {
            return intval($data[0]) + 1;
        }
        
        return 1;
    }
    
    /**
     * Update todo status in CSV
     */
    public function updateTodoStatus($todoId, $newStatus) {
        if (!file_exists($this->csvFile)) {
            return false;
        }
        
        $lines = file($this->csvFile, FILE_IGNORE_NEW_LINES);
        $updated = false;
        
        foreach ($lines as $index => $line) {
            if ($index === 0) continue; // Skip header
            
            $data = str_getcsv($line);
            if (!empty($data[0]) && $data[0] == $todoId) {
                $data[2] = $newStatus; // Update status column
                $lines[$index] = $this->arrayToCsv($data);
                $updated = true;
                break;
            }
        }
        
        if ($updated) {
            file_put_contents($this->csvFile, implode("\n", $lines) . "\n");
            Config::log("Updated todo #$todoId status to: $newStatus");
            
            // Also update in main database
            if (strtolower($newStatus) === 'completed') {
                $this->todoManager->completeTodo($todoId);
            }
            
            return true;
        }
        
        Config::log("Todo #$todoId not found in Google Sheets", 'WARNING');
        return false;
    }
    
    /**
     * Complete a todo
     */
    public function completeTodo($todoId) {
        return $this->updateTodoStatus($todoId, 'Completed');
    }
    
    /**
     * Get todos from CSV with optional status filter
     */
    public function getTodos($statusFilter = null) {
        if (!file_exists($this->csvFile)) {
            return [];
        }
        
        $todos = [];
        $lines = file($this->csvFile, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES);
        
        foreach ($lines as $index => $line) {
            if ($index === 0) continue; // Skip header
            
            $data = str_getcsv($line);
            if (count($data) >= 6) {
                $todo = [
                    'id' => $data[0],
                    'description' => $data[1],
                    'status' => $data[2],
                    'date_added' => $data[3],
                    'due_date' => $data[4] ?? '',
                    'priority' => $data[5] ?? 'Medium',
                    'notes' => $data[6] ?? '',
                    'source' => $data[7] ?? 'sheets'
                ];
                
                if ($statusFilter === null || strcasecmp($todo['status'], $statusFilter) === 0) {
                    $todos[] = $todo;
                }
            }
        }
        
        return $todos;
    }
    
    /**
     * Sync from other sources (WhatsApp, Email, etc.)
     */
    public function syncFromOtherSources($limit = 10) {
        $syncedCount = 0;
        
        // Get recent todos from database that aren't from sheets
        $recentTodos = $this->todoManager->getTodos('pending', null, $limit);
        
        foreach ($recentTodos as $todo) {
            if ($todo['source'] !== 'sheets') {
                // Check if already in CSV
                if (!$this->todoExistsInCSV($todo['description'])) {
                    $notes = "From {$todo['source']}";
                    if (!empty($todo['source_data'])) {
                        $sourceData = $todo['source_data'];
                        if (isset($sourceData['phone_number'])) {
                            $notes .= " ({$sourceData['phone_number']})";
                        } elseif (isset($sourceData['sender_email'])) {
                            $notes .= " ({$sourceData['sender_email']})";
                        }
                    }
                    
                    $this->addTodoToCSVOnly($todo['description'], $todo['priority'], $notes, $todo['source']);
                    $syncedCount++;
                }
            }
        }
        
        Config::log("Synced $syncedCount todos to Google Sheets");
        return $syncedCount;
    }
    
    /**
     * Check if todo exists in CSV (to avoid duplicates)
     */
    private function todoExistsInCSV($description) {
        $todos = $this->getTodos();
        foreach ($todos as $todo) {
            if (strcasecmp($todo['description'], $description) === 0) {
                return true;
            }
        }
        return false;
    }
    
    /**
     * Add todo to CSV only (for syncing)
     */
    private function addTodoToCSVOnly($description, $priority, $notes, $source) {
        $todoId = $this->getNextId();
        $dateAdded = date('Y-m-d');
        
        $row = [
            $todoId,
            $description,
            'Pending',
            $dateAdded,
            '', // due_date
            ucfirst($priority),
            $notes,
            $source
        ];
        
        $fp = fopen($this->csvFile, 'a');
        fputcsv($fp, $row);
        fclose($fp);
    }
    
    /**
     * Export todos to JSON
     */
    public function exportToJson($filename = null) {
        if ($filename === null) {
            $filename = 'sheets_todos_export_' . date('Y-m-d_H-i-s') . '.json';
        }
        
        $todos = $this->getTodos();
        $stats = $this->getStats();
        
        $export = [
            'exported_at' => date('c'),
            'source' => 'google_sheets_csv',
            'stats' => $stats,
            'todos' => $todos
        ];
        
        $filepath = Config::get('backups_dir') . $filename;
        file_put_contents($filepath, json_encode($export, JSON_PRETTY_PRINT));
        
        Config::log("Exported Google Sheets todos to $filename");
        return $filepath;
    }
    
    /**
     * Generate summary statistics
     */
    public function getStats() {
        $todos = $this->getTodos();
        $pending = $this->getTodos('pending');
        $completed = $this->getTodos('completed');
        
        $priorityCount = [];
        foreach ($todos as $todo) {
            $priority = $todo['priority'];
            $priorityCount[$priority] = ($priorityCount[$priority] ?? 0) + 1;
        }
        
        return [
            'total' => count($todos),
            'pending' => count($pending),
            'completed' => count($completed),
            'completion_rate' => count($todos) > 0 ? round((count($completed) / count($todos)) * 100, 1) : 0,
            'priority_breakdown' => $priorityCount
        ];
    }
    
    /**
     * Generate summary report
     */
    public function generateSummaryReport() {
        $stats = $this->getStats();
        $pendingTodos = $this->getTodos('pending');
        
        $report = "📊 GOOGLE SHEETS TODO SUMMARY REPORT\n";
        $report .= "================================\n";
        $report .= "Generated: " . date('Y-m-d H:i:s') . "\n\n";
        
        $report .= "📋 Overview:\n";
        $report .= "• Total Todos: {$stats['total']}\n";
        $report .= "• Pending: {$stats['pending']}\n";
        $report .= "• Completed: {$stats['completed']}\n";
        $report .= "• Completion Rate: {$stats['completion_rate']}% ({$stats['completed']}/{$stats['total']})\n\n";
        
        $report .= "🎯 Priority Breakdown:\n";
        foreach ($stats['priority_breakdown'] as $priority => $count) {
            $report .= "• $priority: $count\n";
        }
        
        $report .= "\n📅 Recent Pending Todos:\n";
        $recentPending = array_slice($pendingTodos, -5); // Last 5 pending
        foreach ($recentPending as $todo) {
            $report .= "• #{$todo['id']}: {$todo['description']}\n";
        }
        
        return $report;
    }
    
    /**
     * Generate mobile-friendly view
     */
    public function getMobileView() {
        $pendingTodos = $this->getTodos('pending');
        
        if (empty($pendingTodos)) {
            return "🎉 No pending todos! You're all caught up!";
        }
        
        $mobileView = "📱 YOUR TODOS (Mobile View)\n";
        $mobileView .= str_repeat("=", 30) . "\n\n";
        
        $displayTodos = array_slice($pendingTodos, 0, 10); // Show top 10
        
        foreach ($displayTodos as $todo) {
            $statusEmoji = $todo['status'] === 'Pending' ? '⏳' : '✅';
            $priorityEmoji = $this->getPriorityEmoji($todo['priority']);
            
            $mobileView .= "$priorityEmoji $statusEmoji #{$todo['id']}\n";
            $mobileView .= "   {$todo['description']}\n";
            
            if (!empty($todo['due_date'])) {
                $mobileView .= "   📅 Due: {$todo['due_date']}\n";
            }
            
            $mobileView .= "\n";
        }
        
        if (count($pendingTodos) > 10) {
            $remaining = count($pendingTodos) - 10;
            $mobileView .= "... and $remaining more todos\n";
        }
        
        return $mobileView;
    }
    
    /**
     * Get priority emoji
     */
    private function getPriorityEmoji($priority) {
        switch (strtolower($priority)) {
            case 'high': return '🔥';
            case 'medium': return '📋';
            case 'low': return '📝';
            default: return '📋';
        }
    }
    
    /**
     * Convert array to CSV string
     */
    private function arrayToCsv($array) {
        $fp = fopen('php://temp', 'r+');
        fputcsv($fp, $array);
        rewind($fp);
        $csv = stream_get_contents($fp);
        fclose($fp);
        return rtrim($csv, "\n");
    }
    
    /**
     * Import todos from CSV file
     */
    public function importFromCSV($csvFilePath) {
        if (!file_exists($csvFilePath)) {
            Config::log("CSV file not found: $csvFilePath", 'ERROR');
            return false;
        }
        
        $importedCount = 0;
        $fp = fopen($csvFilePath, 'r');
        
        // Skip header
        fgetcsv($fp);
        
        while (($data = fgetcsv($fp)) !== false) {
            if (count($data) >= 2) {
                $description = $data[1];
                $priority = $data[5] ?? 'Medium';
                $notes = $data[6] ?? '';
                $dueDate = $data[4] ?? '';
                
                $this->addTodo($description, $dueDate, $priority, $notes, 'import');
                $importedCount++;
            }
        }
        
        fclose($fp);
        Config::log("Imported $importedCount todos from CSV");
        
        return $importedCount;
    }
}

/**
 * Real Google Sheets API Integration (for advanced users)
 * This is a placeholder for actual Google Sheets API implementation
 */
class RealGoogleSheetsIntegration {
    /* 
    To implement real Google Sheets API integration:
    
    1. Install Google API Client Library:
       composer require google/apiclient
    
    2. Set up Google Cloud Project and enable Sheets API
    
    3. Download credentials.json file
    
    4. Implement OAuth2 authentication
    
    5. Use Google_Service_Sheets to interact with sheets
    
    Example code structure:
    
    private function authenticateGoogleSheets() {
        $client = new Google_Client();
        $client->setApplicationName('Todo Management System');
        $client->setScopes(Google_Service_Sheets::SPREADSHEETS);
        $client->setAuthConfig('credentials.json');
        $client->setAccessType('offline');
        
        // Handle authentication flow
        return new Google_Service_Sheets($client);
    }
    
    private function readSheet($spreadsheetId, $range) {
        $service = $this->authenticateGoogleSheets();
        $response = $service->spreadsheets_values->get($spreadsheetId, $range);
        return $response->getValues();
    }
    
    private function writeToSheet($spreadsheetId, $range, $values) {
        $service = $this->authenticateGoogleSheets();
        $body = new Google_Service_Sheets_ValueRange(['values' => $values]);
        $params = ['valueInputOption' => 'RAW'];
        
        return $service->spreadsheets_values->append(
            $spreadsheetId, $range, $body, $params
        );
    }
    */
}
?>

File: old kit/python/google_sheets_integration.py
#!/usr/bin/env python3
"""
Google Sheets Todo Integration
Syncs todos with Google Sheets for mobile-friendly todo management
"""

import os
import json
from datetime import datetime
from env_loader import load_env

# Note: This is a simplified version using CSV for demonstration
# For actual Google Sheets API integration, you would need:
# 1. Google Sheets API credentials
# 2. google-api-python-client library
# 3. OAuth2 authentication setup

class GoogleSheetsTodoManager:
    def __init__(self, sheet_file="google_sheets_todos.csv"):
        load_env()
        self.sheet_file = sheet_file
        self.initialize_sheet()
    
    def initialize_sheet(self):
        """Initialize the CSV file that simulates Google Sheets"""
        if not os.path.exists(self.sheet_file):
            with open(self.sheet_file, 'w') as f:
                # Create header row
                f.write("ID,Todo Item,Status,Date Added,Due Date,Priority,Notes\n")
    
    def add_todo(self, description, due_date="", priority="Medium", notes=""):
        """Add a new todo to the sheet"""
        todo_id = self.get_next_id()
        date_added = datetime.now().strftime("%Y-%m-%d")
        
        with open(self.sheet_file, 'a') as f:
            f.write(f"{todo_id},{description},Pending,{date_added},{due_date},{priority},{notes}\n")
        
        print(f"✅ Added todo #{todo_id}: {description}")
        return todo_id
    
    def get_next_id(self):
        """Get the next available ID"""
        try:
            with open(self.sheet_file, 'r') as f:
                lines = f.readlines()
                if len(lines) <= 1:  # Only header or empty
                    return 1
                # Get the last ID and increment
                last_line = lines[-1].strip()
                if last_line:
                    last_id = int(last_line.split(',')[0])
                    return last_id + 1
        except (FileNotFoundError, ValueError, IndexError):
            pass
        return 1
    
    def complete_todo(self, todo_id):
        """Mark a todo as completed"""
        return self.update_todo_status(todo_id, "Completed")
    
    def update_todo_status(self, todo_id, new_status):
        """Update the status of a todo"""
        try:
            with open(self.sheet_file, 'r') as f:
                lines = f.readlines()
            
            updated = False
            for i, line in enumerate(lines):
                if i == 0:  # Skip header
                    continue
                
                parts = line.strip().split(',')
                if len(parts) >= 3 and parts[0] == str(todo_id):
                    parts[2] = new_status  # Update status column
                    lines[i] = ','.join(parts) + '\n'
                    updated = True
                    break
            
            if updated:
                with open(self.sheet_file, 'w') as f:
                    f.writelines(lines)
                print(f"✅ Updated todo #{todo_id} status to: {new_status}")
                return True
            else:
                print(f"❌ Todo #{todo_id} not found")
                return False
                
        except Exception as e:
            print(f"❌ Error updating todo: {e}")
            return False
    
    def list_todos(self, status_filter=None):
        """List todos with optional status filter"""
        try:
            todos = []
            with open(self.sheet_file, 'r') as f:
                lines = f.readlines()
            
            for i, line in enumerate(lines):
                if i == 0:  # Skip header
                    continue
                
                parts = line.strip().split(',')
                if len(parts) >= 6:
                    todo = {
                        'id': parts[0],
                        'description': parts[1],
                        'status': parts[2],
                        'date_added': parts[3],
                        'due_date': parts[4],
                        'priority': parts[5],
                        'notes': parts[6] if len(parts) > 6 else ""
                    }
                    
                    if status_filter is None or todo['status'].lower() == status_filter.lower():
                        todos.append(todo)
            
            return todos
            
        except Exception as e:
            print(f"❌ Error listing todos: {e}")
            return []
    
    def sync_from_other_sources(self, whatsapp_todos=None, github_issues=None):
        """Sync todos from other sources (WhatsApp, GitHub, etc.)"""
        synced_count = 0
        
        # Sync from WhatsApp todos
        if whatsapp_todos:
            for todo in whatsapp_todos:
                if todo['status'] == 'pending':
                    self.add_todo(
                        description=f"[WhatsApp] {todo['description']}",
                        notes=f"From WhatsApp: {todo['phone_number']}"
                    )
                    synced_count += 1
        
        # Sync from GitHub issues (would require GitHub API)
        if github_issues:
            for issue in github_issues:
                self.add_todo(
                    description=f"[GitHub] {issue['title']}",
                    notes=f"Issue #{issue['number']}"
                )
                synced_count += 1
        
        print(f"📊 Synced {synced_count} todos from other sources")
        return synced_count
    
    def export_to_json(self, filename="sheets_todos_export.json"):
        """Export todos to JSON format"""
        todos = self.list_todos()
        
        with open(filename, 'w') as f:
            json.dump(todos, f, indent=2)
        
        print(f"📤 Exported {len(todos)} todos to {filename}")
        return filename
    
    def generate_summary_report(self):
        """Generate a summary report of todos"""
        all_todos = self.list_todos()
        pending_todos = self.list_todos("pending")
        completed_todos = self.list_todos("completed")
        
        # Priority breakdown
        priority_count = {}
        for todo in all_todos:
            priority = todo['priority']
            priority_count[priority] = priority_count.get(priority, 0) + 1
        
        report = f"""📊 GOOGLE SHEETS TODO SUMMARY REPORT
================================
Generated: {datetime.now().strftime("%Y-%m-%d %H:%M:%S")}

📋 Overview:
• Total Todos: {len(all_todos)}
• Pending: {len(pending_todos)}
• Completed: {len(completed_todos)}
• Completion Rate: {(len(completed_todos)/len(all_todos)*100):.1f}% ({len(completed_todos)}/{len(all_todos)})

🎯 Priority Breakdown:
"""
        
        for priority, count in priority_count.items():
            report += f"• {priority}: {count}\n"
        
        report += f"\n📅 Recent Pending Todos:\n"
        for todo in pending_todos[-5:]:  # Last 5 pending
            report += f"• #{todo['id']}: {todo['description']}\n"
        
        print(report)
        return report
    
    def mobile_friendly_view(self):
        """Generate a mobile-friendly view of todos"""
        pending_todos = self.list_todos("pending")
        
        if not pending_todos:
            return "🎉 No pending todos! You're all caught up!"
        
        mobile_view = "📱 YOUR TODOS (Mobile View)\n"
        mobile_view += "=" * 30 + "\n\n"
        
        for i, todo in enumerate(pending_todos[:10], 1):  # Show top 10
            status_emoji = "⏳" if todo['status'] == "Pending" else "✅"
            priority_emoji = {"High": "🔥", "Medium": "📋", "Low": "📝"}.get(todo['priority'], "📋")
            
            mobile_view += f"{priority_emoji} {status_emoji} #{todo['id']}\n"
            mobile_view += f"   {todo['description']}\n"
            
            if todo['due_date']:
                mobile_view += f"   📅 Due: {todo['due_date']}\n"
            
            mobile_view += "\n"
        
        if len(pending_todos) > 10:
            mobile_view += f"... and {len(pending_todos) - 10} more todos\n"
        
        return mobile_view

# Actual Google Sheets API Integration (commented out - requires setup)
"""
To use real Google Sheets API, you would need:

1. Install dependencies:
   pip install google-api-python-client google-auth google-auth-oauthlib

2. Set up Google Cloud Project and enable Sheets API

3. Download credentials.json file

4. Use this code structure:

from googleapiclient.discovery import build
from google.oauth2.credentials import Credentials
from google_auth_oauthlib.flow import InstalledAppFlow
from google.auth.transport.requests import Request

class RealGoogleSheetsManager:
    def __init__(self, spreadsheet_id, range_name="Sheet1!A:G"):
        self.spreadsheet_id = spreadsheet_id
        self.range_name = range_name
        self.service = self.authenticate()
    
    def authenticate(self):
        creds = None
        # Authentication logic here
        service = build('sheets', 'v4', credentials=creds)
        return service
    
    def read_sheet(self):
        result = self.service.spreadsheets().values().get(
            spreadsheetId=self.spreadsheet_id,
            range=self.range_name
        ).execute()
        return result.get('values', [])
    
    def write_to_sheet(self, values):
        body = {'values': values}
        result = self.service.spreadsheets().values().append(
            spreadsheetId=self.spreadsheet_id,
            range=self.range_name,
            valueInputOption='RAW',
            body=body
        ).execute()
        return result
"""

def main():
    """Test the Google Sheets integration"""
    print("🧪 Testing Google Sheets Todo Integration")
    print("=" * 40)
    
    manager = GoogleSheetsTodoManager()
    
    # Add some test todos
    manager.add_todo("Review quarterly budget", "2025-02-15", "High", "Finance review")
    manager.add_todo("Call investor meeting", "2025-02-10", "High", "Important call")
    manager.add_todo("Update website content", "2025-02-20", "Medium", "Marketing task")
    manager.add_todo("Plan team outing", "2025-02-25", "Low", "HR task")
    
    print("\n📋 All Todos:")
    todos = manager.list_todos()
    for todo in todos:
        print(f"  #{todo['id']}: {todo['description']} ({todo['status']})")
    
    # Complete a todo
    print(f"\n✅ Completing todo #2...")
    manager.complete_todo(2)
    
    # Show mobile view
    print(f"\n{manager.mobile_friendly_view()}")
    
    # Generate report
    print(f"\n{manager.generate_summary_report()}")
    
    # Export to JSON
    manager.export_to_json()

if __name__ == "__main__":
    main()